File: server/sonar-web/src/main/js/helpers/testUtils.js
/*
 * SonarQube
 * Copyright (C) 2009-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
export const mockEvent = {
  target: { blur() {} },
  currentTarget: { blur() {} },
  preventDefault() {},
  stopPropagation() {}
};

export const click = (element, event = {}) => element.simulate('click', { ...mockEvent, ...event });

export const clickOutside = (event = {}) => {
  const dispatchedEvent = new MouseEvent('click', event);
  window.dispatchEvent(dispatchedEvent);
};

export const submit = element =>
  element.simulate('submit', {
    preventDefault() {}
  });

export const change = (element, value) =>
  element.simulate('change', {
    target: { value },
    currentTarget: { value }
  });

export const keydown = keyCode => {
  const event = new KeyboardEvent('keydown', { keyCode });
  document.dispatchEvent(event);
};

export const elementKeydown = (element, keyCode) => {
  element.simulate('keydown', {
    currentTarget: { element },
    keyCode,
    preventDefault() {}
  });
};

File: sonar-plugin-api/src/test/java/org/sonar/api/utils/TestUtils.java
/*
 * SonarQube
 * Copyright (C) 2009-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
package org.sonar.api.utils;

import java.lang.reflect.Constructor;
import java.lang.reflect.Modifier;

/**
 * Utilities for unit tests
 *
 * @since 2.2
 */
public final class TestUtils {

  private TestUtils() {
  }

  /**
   * Asserts that all constructors are private, usually for helper classes with
   * only static methods. If a constructor does not have any parameters, then
   * it's instantiated.
   */
  public static boolean hasOnlyPrivateConstructors(Class clazz) {
    boolean ok = true;
    for (Constructor constructor : clazz.getDeclaredConstructors()) {
      ok &= Modifier.isPrivate(constructor.getModifiers());
      if (constructor.getParameterTypes().length == 0) {
        constructor.setAccessible(true);
        try {
          constructor.newInstance();
        } catch (Exception e) {
          throw new IllegalStateException(String.format("Fail to instantiate %s", clazz), e);
        }
      }
    }
    return ok;
  }
}
